crates/ra_prof/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+use std::mem;
+use std::io::{stderr, StderrLock, Write};
+use std::iter::repeat;
+use std::collections::{HashSet};
+use std::default::Default;
+use std::iter::FromIterator;
+use std::sync::RwLock;
+
+use lazy_static::lazy_static;
+
+type Message = (usize, u64, String);
+
+pub struct Profiler {
+    desc: String,
+}
+
+pub struct Filter {
+    depth: usize,
+    allowed: Vec<String>,
+}
+
+struct ProfileStack {
+    starts: Vec<time::Instant>,
+    messages: Vec<Message>,
+    filter_data: FilterData,
+}
+
+impl ProfileStack {
+    fn new() -> ProfileStack {
+        ProfileStack { starts: Vec::new(), messages: Vec::new(), filter_data: Default::default() }
+    }
+}
+
+#[derive(Default)]
+struct FilterData {
+    depth: usize,
+    version: usize,
+    allowed: HashSet<String>,
+}
+
+impl Clone for FilterData {
+    fn clone(&self) -> FilterData {
+        let set = HashSet::from_iter(self.allowed.iter().cloned());
+        FilterData { depth: self.depth, allowed: set, version: self.version }
+    }
+}
+
+lazy_static! {
+    static ref FILTER: RwLock<FilterData> = RwLock::new(Default::default());
+}
+
+thread_local!(static PROFILE_STACK: RefCell<ProfileStack> = RefCell::new(ProfileStack::new()));
+
+pub fn set_filter(f: Filter) {
+    let mut old = FILTER.write().unwrap();
+    let set = HashSet::from_iter(f.allowed.iter().cloned());
+    let filter_data = FilterData { depth: f.depth, allowed: set, version: old.version + 1 };
+    *old = filter_data;
+}
+
+pub fn profile<T: fmt::Display>(desc: T) -> Profiler {
+    PROFILE_STACK.with(|stack| {
+        let mut stack = stack.borrow_mut();
+        if stack.starts.len() == 0 {
+            match FILTER.try_read() {
+                Ok(f) => {
+                    if f.version > stack.filter_data.version {
+                        stack.filter_data = f.clone();
+                    }
+                }
+                Err(_) => (),
+            };
+        }
+        let desc_str = desc.to_string();
+        if desc_str.is_empty() {
+            Profiler { desc: desc_str }
+        } else if stack.starts.len() < stack.filter_data.depth
+            && stack.filter_data.allowed.contains(&desc_str)
+        {
+            stack.starts.push(time::Instant::now());
+            Profiler { desc: desc_str }
+        } else {
+            Profiler { desc: String::new() }
+        }
+    })
+}
+
+impl Drop for Profiler {
+    fn drop(&mut self) {
+        if self.desc.is_empty() {
+            return;
+        }
+        PROFILE_STACK.with(|stack| {
+            let mut stack = stack.borrow_mut();
+            let start = stack.starts.pop().unwrap();
+            let duration = start.elapsed();
+            let duration_ms = duration.as_secs() * 1000 + u64::from(duration.subsec_millis());
+            let stack_len = stack.starts.len();
+            let msg = (stack_len, duration_ms, mem::replace(&mut self.desc, String::new()));
+            stack.messages.push(msg);
+            if stack_len == 0 {
```

Comment:
We are focusig on live analysis, so it makes sense to print the data as it comes out, and not when we reach zero stack.

Reply:
I don't really get what to do with profile nesting in case of real time output.
If we have two nested profiles we can print two at once like it is right now, or we can print inner profile when it drops and both nested profiles when outer drops. In latter case inner profile will be printed twice. Did you expect the second case?

Reply:
Hm, yeah, I guess what I am proposing doesn't make any sense :) 

I somehow thought that we reach zero stack only when we terminate the whole process, but of course this is not the case. Current impl is good,  sorry for being confusing!